src/Features/Core/Portable/CodeRefactorings/AddMissingImports/AbstractAddMissingImportsFeatureService.cs

```diff
@@ -0,0 +1,160 @@
+            }
+
+            // Apply those fixes to the document.
+            var newDocument = await ApplyFixesAsync(document, usableFixes, cancellationToken).ConfigureAwait(false);
+            return newDocument.Project;
+        }
+
+        private async Task<ImmutableArray<Diagnostic>> GetDiagnosticsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        {
+            var project = document.Project;
+
+            var diagnosticData = await _diagnosticAnalyzerService.GetDiagnosticsForSpanAsync(document, textSpan, cancellationToken: cancellationToken).ConfigureAwait(false);
+            var getDiagnosticTasks = diagnosticData
+                .Where(datum => FixableDiagnosticIds.Contains(datum.Id))
+                .Select(datum => datum.ToDiagnosticAsync(project, cancellationToken));
+
+            await Task.WhenAll(getDiagnosticTasks).ConfigureAwait(false);
+
+            var diagnostics = ArrayBuilder<Diagnostic>.GetInstance();
+            foreach (var getDiagnosticTask in getDiagnosticTasks)
+            {
+                var diagnostic = await getDiagnosticTask.ConfigureAwait(false);
+                diagnostics.Add(diagnostic);
+            }
+
+            return diagnostics.ToImmutableAndFree()
+                .Sort((d1, d2) => d1.Location.SourceSpan.Start - d2.Location.SourceSpan.Start);
+        }
+
+        private async Task<ImmutableArray<AddImportFixData>> GetUnambiguousFixesAsync(Document document, ImmutableArray<Diagnostic> diagnostics, CancellationToken cancellationToken)
+        {
+            var solution = document.Project.Solution;
+            var symbolSearchService = solution.Workspace.Services.GetService<ISymbolSearchService>();
+            // Since we are not currently considering NuGet packages, pass an empty array
+            var packageSources = ImmutableArray<PackageSource>.Empty;
+            var addImportService = document.GetLanguageService<IAddImportFeatureService>();
+
+            var getFixesForSpanTasks = diagnostics
+                .GroupBy(diagnostic => diagnostic.Location.SourceSpan)
+                .Select(diagnosticsForSourceSpan => addImportService
+                    .GetFixesForDiagnosticsAsync(document, diagnosticsForSourceSpan.Key, diagnosticsForSourceSpan.AsImmutable(), symbolSearchService, searchReferenceAssemblies: true, packageSources, cancellationToken));
+
+            await Task.WhenAll(getFixesForSpanTasks).ConfigureAwait(false);
+
+            var fixes = ArrayBuilder<AddImportFixData>.GetInstance();
+            foreach (var getFixesForSpanTask in getFixesForSpanTasks)
+            {
+                var fixesForSpan = await getFixesForSpanTask.ConfigureAwait(false);
+
+                foreach (var fixForSpan in fixesForSpan)
+                {
+                    // If there is more than one fix, then we will leave it for the user to manually apply.
+                    if (fixForSpan.Fixes.Length == 1)
```

Comment:
the naming here is a bit strange.  its called a "fixForSpan" but inside contains "fixes".  so it really seems like it represents something different